vllm/distributed/kv_transfer/kv_connector/v1/cpu_connector.py

```diff
@@ -0,0 +1,936 @@
+                self._kv_receiver.free_request(p_req_id)
+            self._inflight_h2d_requests.clear()
+
+    @_lmcache_nvtx_annotate
+    def save_kv_layer(self, layer_name: str, kv_layer: torch.Tensor,
+                      attn_metadata: "AttentionMetadata", **kwargs) -> None:
+        """
+        Start saving a layer of KV cache from vLLM's paged buffer 
+        to the connector. This is called from within attention layer to
+        enable async copying during execution.
+
+        Args:
+            layer_name (str): the name of the layer.
+            kv_layer (torch.Tensor): the paged KV buffer of the current 
+                layer in vLLM.
+            attn_metadata (AttentionMetadata): the attention metadata.
+            **kwargs: additional arguments for the save operation.
+        """
+        if self.kv_role == "kv_consumer":
+            # decoder side
+            return
+
+        meta = self._get_connector_metadata()
+        assert isinstance(meta, CPUConnectorMetadata), \
+                "Connector metadata is not of type CPUConnectorMetadata"
+        assert self._kv_sender is not None
+
+        for prefill_req in meta.prefill_meta:
+            # Create a source spec with serializable types
+            source_spec = SourceSpec(
+                request_id=prefill_req.req_id,
+                layer_id=self._get_layer_id(layer_name),
+                start=prefill_req.token_range.start,
+                stop=prefill_req.token_range.stop,
+                shape=tuple(self._get_kv_shape(len(
+                    prefill_req.blocks_to_save))),
+                dtype_str=str(kv_layer.dtype).split('.')
+                [-1],  # Convert torch.float32 -> "float32"
+                num_all_tokens=prefill_req.num_all_tokens,
+            )
+
+            # Create a destination spec
+            dest_spec = DestinationSpec(
+                rank=get_tensor_model_parallel_rank(),
+                host=self._host,
+                base_port=self._port,
+            )
+
+            # Create the send task
+            with self._kv_sender_lock:
+                task = self._kv_sender.create_send_task(
+                    source_spec=source_spec,
+                    destination_spec=dest_spec,
+                )
+            assert isinstance(task, NixlSendTask), \
+                    "Send task is not of type NixlSendTask"
+
+            # Start copying the data to the CPU buffer
+            buffer = task.tensor
+            self._cuda_stream.wait_stream(torch.cuda.current_stream())
+            with torch.cuda.stream(self._cuda_stream):
+                # Copy the data from the GPU to the CPU buffer page by page
+                rng = nvtx.start_range("d2h_page_copy")
+                d2h_page_copy(src_layer=kv_layer,
+                              dst_buffer=buffer,
+                              block_ids=prefill_req.blocks_to_save)
+                nvtx.end_range(rng)
```

Comment:
Can you add some comments about what nvtx is and why its needed? This is an unfamiliar API so I think it warrants some additional commentary